File: src/main/java/com/matag/game/turn/Turn.java
package com.matag.game.turn;

import com.matag.game.cardinstance.CardInstance;
import com.matag.game.status.GameStatus;
import lombok.EqualsAndHashCode;
import lombok.Getter;
import lombok.Setter;
import lombok.ToString;
import org.springframework.context.annotation.Scope;
import org.springframework.stereotype.Component;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

@ToString
@EqualsAndHashCode
@Getter
@Setter
@Component
@Scope("prototype")
public class Turn {
  private int turnNumber;
  private String currentTurnPlayer;
  private String currentPhase;
  private String currentPhaseActivePlayer;
  private List<CardInstance> cardsPlayedWithinTurn = new ArrayList<>();
  private Map<Integer, List<String>> lastManaPaid = new HashMap<>();
  private String inputRequiredAction;
  private String inputRequiredActionParameter;
  private String winner;

  public void init(String playerName) {
    this.turnNumber = 1;
    this.currentPhase = "UP";
    this.currentTurnPlayer = playerName;
    this.currentPhaseActivePlayer = playerName;
  }

  public void increaseTurnNumber() {
    turnNumber++;
  }

  public void addCardToCardsPlayedWithinTurn(CardInstance cardInstance) {
    cardsPlayedWithinTurn.add(cardInstance);
  }

  public boolean isEnded() {
    return winner != null;
  }

  public void passPriority(GameStatus gameStatus) {
    if (currentPhaseActivePlayer.equals(gameStatus.getPlayer1().getName())) {
      currentPhaseActivePlayer = gameStatus.getPlayer2().getName();
    } else {
      currentPhaseActivePlayer = gameStatus.getPlayer1().getName();
    }
  }
}

File: src/main/kotlin/com/matag/game/turn/Turn.kt
package com.matag.game.turn

import com.matag.game.cardinstance.CardInstance
import com.matag.game.status.GameStatus
import org.springframework.context.annotation.Scope
import org.springframework.stereotype.Component

@Component
@Scope("prototype")
class Turn {
    var turnNumber = 0
    var currentTurnPlayer: String? = null
    var currentPhase: String? = null
    var currentPhaseActivePlayer: String? = null
    var cardsPlayedWithinTurn: List<CardInstance> = listOf()
    var lastManaPaid: Map<Int, List<String>> = mapOf()
    var inputRequiredAction: String? = null
    var inputRequiredActionParameter: String? = null
    var winner: String? = null

    fun init(playerName: String) {
        this.turnNumber = 1
        this.currentPhase = "UP"
        this.currentTurnPlayer = playerName
        this.currentPhaseActivePlayer = playerName
    }

    fun increaseTurnNumber() {
        turnNumber++
    }

    fun addCardToCardsPlayedWithinTurn(cardInstance: CardInstance) {
        cardsPlayedWithinTurn += cardInstance
    }

    fun isEnded(): Boolean =
        winner != null

    fun passPriority(gameStatus: GameStatus) {
        if (currentPhaseActivePlayer == gameStatus.player1!!.name) {
            currentPhaseActivePlayer = gameStatus.player2!!.name
        } else {
            currentPhaseActivePlayer = gameStatus.player1!!.name
        }
    }
}